File: plugins/git4idea/tests/git4idea/crlf/GitCrlfProblemsDetectorTest.java
package git4idea.crlf;

import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vcs.Executor;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.containers.ContainerUtil;
import git4idea.test.GitSingleRepoTest;

import java.io.File;
import java.io.IOException;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import static java.util.Arrays.asList;

public class GitCrlfProblemsDetectorTest extends GitSingleRepoTest {
  public void setUp() throws Exception {
    try {
      super.setUp();
    }
    catch (Exception e) {
      super.tearDown();
      throw e;
    }
    Executor.cd(projectRoot);
    git("config core.autocrlf false");
  }

  public void test_no_warning_if_autocrlf_is_true() throws IOException {
    git("config core.autocrlf true");
    assertFalse("No warning should be done if core.autocrlf is true", detect("temp").shouldWarn());
  }

  public void test_no_warning_if_autocrlf_is_input() throws IOException {
    git("config core.autocrlf input");
    assertFalse("No warning should be done if core.autocrlf is input", detect("temp").shouldWarn());
  }

  public void test_no_warning_if_no_files_with_CRLF() throws IOException {
    createFile("temp", "Unix file\nNice separators\nOnly LF\n");
    assertFalse("No warning should be done if all files are LFs", detect("temp").shouldWarn());
  }

  public void test_no_warning_if_file_with_CRLF_but_text_is_set() throws IOException {
    gitattributes("*       text=auto");
    createCrlfFile("win");
    assertFalse("No warning should be done if the file has a text attribute", detect("win").shouldWarn());
  }

  public void test_no_warning_if_file_with_CRLF_but_crlf_is_set() throws IOException {
    gitattributes("win       crlf");
    createCrlfFile("win");
    assertFalse("No warning should be done if the file has a crlf attribute", detect("win").shouldWarn());
  }

  public void test_no_warning_if_file_with_CRLF_but_crlf_is_explicitly_unset() throws IOException {
    gitattributes("win       -crlf");
    createCrlfFile("win");
    assertFalse("No warning should be done if the file has an explicitly unset crlf attribute", detect("win").shouldWarn());
  }

  public void test_no_warning_if_file_with_CRLF_but_crlf_is_set_to_input() throws IOException {
    gitattributes("wi*       crlf=input");
    createCrlfFile("win");
    assertFalse("No warning should be done if the file has a crlf attribute", detect("win").shouldWarn());
  }

  public void test_warning_if_file_with_CRLF_no_attrs_autocrlf_is_false() throws IOException {
    createCrlfFile("win");
    assertTrue("Warning should be done if the file has CRLFs inside, and no explicit attributes", detect("win").shouldWarn());
  }

  public void test_warning_if_various_files_with_various_attributes_one_doesnt_match() throws IOException {
    gitattributes("\nwin1 text crlf diff\nwin2 -text\nwin3 text=auto\nwin4 crlf\nwin5 -crlf\nwin6 crlf=input\n");

    createFile("unix", "Unix file\nNice separators\nOnly LF\n");
    createCrlfFile("win1");
    createCrlfFile("win2");
    createCrlfFile("win3");
    createCrlfFile("src/win4");
    createCrlfFile("src/win5");
    createCrlfFile("src/win6");
    createCrlfFile("src/win7");

    List<VirtualFile> files = ContainerUtil.map(asList("unix", "win1", "win2", "win3", "src/win4", "src/win5", "src/win6", "src/win7"),
                                                s -> VfsUtil.findFileByIoFile(new File(projectRoot.getPath(), s), true));
    assertTrue("Warning should be done, since one of the files has CRLFs and no related attributes",
               GitCrlfProblemsDetector.detect(myProject, git, files).shouldWarn());
  }

  private void gitattributes(String content) throws IOException {
    createFile(".gitattributes", content);
  }

  private GitCrlfProblemsDetector detect(String relPath) throws IOException {
    return detect(VfsUtil.findFileByIoFile(createFile(relPath), true));
  }

  private GitCrlfProblemsDetector detect(VirtualFile file) {
    return GitCrlfProblemsDetector.detect(myProject, git, Collections.singleton(file));
  }

  private void createCrlfFile(String relPath) throws IOException {
    createFile(relPath, "Windows file\r\nBad separators\r\nCRLF in action\r\n");
  }

  private void createFile(String relPath, String content) throws IOException {
    File file = createFile(relPath);
    FileUtil.appendToFile(file, content);
  }

  private File createFile(String relPath) throws IOException {
    List<String> split = StringUtil.split(relPath, "/");
    File parent = new File(projectRoot.getPath());
    for (Iterator<String> it = split.iterator(); it.hasNext(); ) {
      String item = it.next();
      File file = new File(parent, item);
      if (it.hasNext()) {
        parent = file;
        parent.mkdir();
      }
      else {
        file.createNewFile();
        return file;
      }
    }

    return parent;
  }
}

File: plugins/git4idea/tests/git4idea/crlf/GitCrlfProblemsDetectorTest.kt
// Copyright 2000-2026 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package git4idea.crlf

import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.testFramework.junit5.TestApplication
import git4idea.test.GitSingleRepoContext
import git4idea.test.git
import git4idea.test.gitSingleRepoContextFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import kotlin.io.path.writeText

@TestApplication
class GitCrlfProblemsDetectorTest {
  private val fixture = gitSingleRepoContextFixture()
  private val context: GitSingleRepoContext get() = fixture.get()

  @BeforeEach
  fun setUp(): Unit = with(context) {
    git("config core.autocrlf false")
  }

  @Test
  fun `test no warning if autocrlf is true`(): Unit = with(context) {
    git("config core.autocrlf true")
    assertThat(detect("temp").shouldWarn())
      .describedAs("No warning should be done if core.autocrlf is true")
      .isFalse()
  }

  @Test
  fun `test no warning if autocrlf is input`(): Unit = with(context) {
    git("config core.autocrlf input")
    assertThat(detect("temp").shouldWarn())
      .describedAs("No warning should be done if core.autocrlf is input")
      .isFalse()
  }

  @Test
  fun `test no warning if no files with CRLF`(): Unit = with(context) {
    createFile("temp", "Unix file\nNice separators\nOnly LF\n")
    assertThat(detect("temp").shouldWarn())
      .describedAs("No warning should be done if all files are LFs")
      .isFalse()
  }

  @Test
  fun `test no warning if file with CRLF but text is set`(): Unit = with(context) {
    gitattributes("*       text=auto")
    createCrlfFile("win")
    assertThat(detect("win").shouldWarn())
      .describedAs("No warning should be done if the file has a text attribute")
      .isFalse()
  }

  @Test
  fun `test no warning if file with CRLF but crlf is set`(): Unit = with(context) {
    gitattributes("win       crlf")
    createCrlfFile("win")
    assertThat(detect("win").shouldWarn())
      .describedAs("No warning should be done if the file has a crlf attribute")
      .isFalse()
  }

  @Test
  fun `test no warning if file with CRLF but crlf is explicitly unset`(): Unit = with(context) {
    gitattributes("win       -crlf")
    createCrlfFile("win")
    assertThat(detect("win").shouldWarn())
      .describedAs("No warning should be done if the file has an explicitly unset crlf attribute")
      .isFalse()
  }

  @Test
  fun `test no warning if file with CRLF but crlf is set to input`(): Unit = with(context) {
    gitattributes("wi*       crlf=input")
    createCrlfFile("win")
    assertThat(detect("win").shouldWarn())
      .describedAs("No warning should be done if the file has a crlf attribute")
      .isFalse()
  }

  @Test
  fun `test warning if file with CRLF no attrs autocrlf is false`(): Unit = with(context) {
    createCrlfFile("win")
    assertThat(detect("win").shouldWarn())
      .describedAs("Warning should be done if the file has CRLFs inside, and no explicit attributes")
      .isTrue()
  }

  @Test
  fun `test warning if various files with various attributes one does not match`(): Unit = with(context) {
    gitattributes("\nwin1 text crlf diff\nwin2 -text\nwin3 text=auto\nwin4 crlf\nwin5 -crlf\nwin6 crlf=input\n")

    createFile("unix", "Unix file\nNice separators\nOnly LF\n")
    createCrlfFile("win1")
    createCrlfFile("win2")
    createCrlfFile("win3")
    createCrlfFile("src/win4")
    createCrlfFile("src/win5")
    createCrlfFile("src/win6")
    createCrlfFile("src/win7")

    val files = listOf("unix", "win1", "win2", "win3", "src/win4", "src/win5", "src/win6", "src/win7")
      .map { findFile(projectNioRoot.resolve(it)) }
    assertThat(GitCrlfProblemsDetector.detect(project, git, files).shouldWarn())
      .describedAs("Warning should be done, since one of the files has CRLFs and no related attributes")
      .isTrue()
  }

  private fun GitSingleRepoContext.gitattributes(content: String) {
    createFile(".gitattributes", content)
  }

  private fun GitSingleRepoContext.detect(relativePath: String): GitCrlfProblemsDetector = detect(findFile(createFile(relativePath)))

  private fun GitSingleRepoContext.detect(file: VirtualFile): GitCrlfProblemsDetector =
    GitCrlfProblemsDetector.detect(project, git, listOf(file))

  private fun GitSingleRepoContext.createCrlfFile(relativePath: String) {
    createFile(relativePath, "Windows file\r\nBad separators\r\nCRLF in action\r\n")
  }

  private fun GitSingleRepoContext.createFile(relativePath: String, content: String = ""): Path {
    val file = projectNioRoot.resolve(relativePath)
    Files.createDirectories(file.parent)
    if (Files.notExists(file)) Files.createFile(file)
    file.writeText(content, Charsets.UTF_8, StandardOpenOption.APPEND)
    return file
  }

  private fun findFile(file: Path): VirtualFile =
    checkNotNull(VfsUtil.findFile(file, true))
}
